File: app/src/main/java/com/iceteaviet/fastfoodfinder/model/Store/UserStoreList.java
package com.iceteaviet.fastfoodfinder.model.Store;

import android.os.Parcel;
import android.os.Parcelable;

import com.google.firebase.database.PropertyName;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by nhoxb on 12/6/2016.
 */
public class UserStoreList implements Parcelable {
    public static final int ID_SAVED = 0;
    public static final int ID_FAVOURITE = 1;
    public static final int ID_CHECKED_IN = 2;
    public static final Creator<UserStoreList> CREATOR = new Creator<UserStoreList>() {
        @Override
        public UserStoreList createFromParcel(Parcel in) {
            return new UserStoreList(in);
        }

        @Override
        public UserStoreList[] newArray(int size) {
            return new UserStoreList[size];
        }
    };
    private int id;
    private String listName;
    private int iconId;
    @PropertyName("storeIdList")
    private List<Integer> storeIdList;

    protected UserStoreList(Parcel in) {
        id = in.readInt();
        listName = in.readString();
        iconId = in.readInt();
        storeIdList = new ArrayList<>();
        in.readList(storeIdList, Integer.class.getClassLoader());
    }

    public UserStoreList() {
    }


    public UserStoreList(int id, List<Integer> storeIdList, int iconId, String listName) {
        this.id = id;
        this.iconId = iconId;
        this.listName = listName;

        if (storeIdList == null)
            this.storeIdList = new ArrayList<>();
        else
            this.storeIdList = storeIdList;
    }

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public List<Integer> getStoreIdList() {
        return storeIdList;
    }

    @PropertyName("storeIdList")
    public void setStoreIdList(List<Integer> storeIdList) {
        this.storeIdList = storeIdList;
    }

    public void addStore(Store store) {
        storeIdList.add(store.getId());
    }

    public void addStore(int storeId) {
        storeIdList.add(storeId);
    }

    public void removeStore(int storeId) {
        storeIdList.remove(Integer.valueOf(storeId));
    }

    public String getListName() {
        return listName;
    }

    public void setListName(String listName) {
        this.listName = listName;
    }

    public int getIconId() {
        return iconId;
    }

    public void setIconId(int iconId) {
        this.iconId = iconId;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel parcel, int i) {
        parcel.writeInt(id);
        parcel.writeString(listName);
        parcel.writeInt(iconId);
        parcel.writeList(storeIdList);
    }
}

File: app/src/main/kotlin/com/iceteaviet/fastfoodfinder/data/remote/user/model/UserStoreList.kt
package com.iceteaviet.fastfoodfinder.data.remote.user.model

import android.os.Parcel
import android.os.Parcelable
import com.google.firebase.database.PropertyName
import com.iceteaviet.fastfoodfinder.data.local.user.model.UserStoreListEntity
import com.iceteaviet.fastfoodfinder.data.remote.store.model.Store
import com.iceteaviet.fastfoodfinder.utils.realmListToList

/**
 * Created by Genius Doan on 12/6/2016.
 */
class UserStoreList : Parcelable {
    var id: Int = -1
    var listName: String = ""
    var iconId: Int = -1
    private var storeIdList: MutableList<Int> = ArrayList()

    protected constructor(`in`: Parcel) {
        id = `in`.readInt()
        listName = `in`.readString()
        iconId = `in`.readInt()
        storeIdList = ArrayList()
        `in`.readList(storeIdList, Int::class.java.classLoader)
    }

    constructor()


    constructor(id: Int, storeIdList: MutableList<Int>?, iconId: Int, listName: String) {
        this.id = id
        this.iconId = iconId
        this.listName = listName

        if (storeIdList == null)
            this.storeIdList = ArrayList()
        else
            this.storeIdList = storeIdList
    }

    constructor(userStoreListEntity: UserStoreListEntity) {
        this.id = userStoreListEntity.id
        this.iconId = userStoreListEntity.iconId
        this.listName = userStoreListEntity.listName
        this.storeIdList = realmListToList(userStoreListEntity.storeIdList)
    }

    fun getStoreIdList(): MutableList<Int> {
        return storeIdList
    }

    @PropertyName("storeIdList")
    fun setStoreIdList(storeIdList: MutableList<Int>) {
        this.storeIdList = storeIdList
    }

    fun addStore(store: Store) {
        storeIdList.add(store.id)
    }

    fun addStore(storeId: Int) {
        storeIdList.add(storeId)
    }

    fun removeStore(storeId: Int) {
        storeIdList.remove(Integer.valueOf(storeId))
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(parcel: Parcel, i: Int) {
        parcel.writeInt(id)
        parcel.writeString(listName)
        parcel.writeInt(iconId)
        parcel.writeList(storeIdList)
    }

    companion object CREATOR : Parcelable.Creator<UserStoreList> {
        const val ID_SAVED = 0
        const val ID_FAVOURITE = 1

        override fun createFromParcel(parcel: Parcel): UserStoreList {
            return UserStoreList(parcel)
        }

        override fun newArray(size: Int): Array<UserStoreList?> {
            return arrayOfNulls(size)
        }
    }
}
